content/blog/2022-05-06-end-to-end-computer-vision-api-part-3-remote-exp-ci-cd.md

```diff
@@ -0,0 +1,292 @@
+  DVCLive and Iterative Studio.
+
+descriptionLong: |
+
+  This is the third and last part of this series of posts on how to build an end-to-end Computer Vision solution according to the best available MLOps practices. 
+
+  In part 1, we talked about exploratory work in Jupyter Notebooks; versioning data in remote storage with DVC; and refactoring the code from Jupyter Notebooks into DVC pipeline stages.
+
+  Part 2 talked about the process of managing experiments with DVC pipelines, DVCLive and Iterative Studio.
+
+picture: <TBD>
+author: alex_kim
+commentsUrl: https://discuss.dvc.org/t/end-to-end-computer-vision/1178
+tags:
+  - Computer Vision
+  - DVC
+  - CML
+  - Studio
+  - CI/CD
+  - Experiment Tracking
+---
+
+### Introduction
+
+In this post, we’ll focus on leveraging cloud infrastructure with CML; how to
+enable automatic reporting (with metrics, plots, and other visuals) for PRs; and
+the eventual deployment process.
+
+### Leveraging Cloud Resources with CI/CD and CML
+
+If you ever wondered if there's an easy way to quickly and easily:
+
+1. provision a powerful VM in the cloud (training CV models often requires
+   powerful GPUs rarely available on local machines)
+2. submit your ML training job to it
+3. get the results back
+4. automatically shut down the VM without having to worry about excessive cloud
+   bills
+
+then the answer is yes if you use the [CML library](https://cml.dev/) in
+combination with CI/CD tools like GitHub Actions or GitLab CI/CD.
+
+![Continuous Integration and Deployment for Machine Learning](https://dvc.org/static/300c88b3b1b5f65753629d661cc916e5/2e49e/cicd4ml.png)
+
+We've configured three
+[workflow files](https://github.com/iterative/magnetic-tiles-defect/tree/main/.github/workflows)
+for GitHub Actions, each of which corresponds to a particular stage depending on
+the project's lifecycle we are in:
+
+1. [Workflow for experimentation and hyperparameter tuning](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/1-experiment.yaml)
+
+   ![Workflow for experimentation and hyperparameter tuning](/uploads/images/2022-05-06/workflow_exp.png '=800')
+
+   In this stage, we'll be working on an experimentation git branch:
+   experimenting with data preprocessing, changing model architecture, tuning
+   hyperparameters, etc. Once we think our experiment is ready to be run, we'll
+   push our changes to a remote repository (in this case, GitHub). This push
+   will trigger a CI/CD job in GitHub Actions, which in turn will:
+
+   1. provision an EC2 virtual machine with a GPU in AWS:
+
+      ```yaml
+      ...
+      - name: deploy
+              env:
+                REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
+              run: |
+                cml runner \
+                    --cloud=aws \
+                    --cloud-region=us-east-1 \
+                    --cloud-type=g4dn.xlarge \
+                    --labels=cml-runner
```

Comment:
- indentation
- reword a little to match https://cml.dev/doc/self-hosted-runners

```suggestion
      - name: Deploy runner on AWS EC2
        env:
          REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
        run: |
          cml runner \
              --cloud=aws \
              --cloud-region=us-east-1 \
              --cloud-type=g4dn.xlarge \
              --labels=cml-runner
```